[develop] Use templates for METplus config files

gsketefian opened the pull request:
## DESCRIPTION OF CHANGES: 
<!-- One or more paragraphs describing the problem, solution, and required changes. -->
This PR converts the METplus configuration files into Jinja templates form which the config files needed by the vx tasks can be generated.  These config files are placed in the experiment directory.  This prevents accidental changes to the METplus config files when multiple experiments are running simultaneously.  This PR also reduces the number of J-jobs and ex-scripts used by the vx tasks by combining ones that are similar.  Details:
1) Convert the METplus configuration files into Jinja templates.  In the process, to avoid repetition, combine multiple files for APCP hours greater than 01 (03, 06, and 24) into a single file having the string `_APCPgt01h` in its name.
2) Modify the ex-scripts so that they use the `fill_jinja_template.py` script to generate the METplus config files from the templates.
3) Combine similar j-jobs and ex-scripts into ones that can be used by multiple tasks, e.g. combine the ex-script that calls METplus's `GridStat` and `PointStat` tools for deterministic verification into one ex-script with the string `gridstat_or_pointstat` in its name.
4) Modify the ROCOTO template XML to use the new J-job and ex-script names and pass in several new variables to the vx tasks (`OBTYPE`, `ACCUM_HH` (if necessary), and `MET_TOOL`; also, remove `FHR`).
5) Add a new WE2E test (`MET_ensemble_verification_only_vx`) that tests use of staged obs and forecast files to run ensemble verification. 
6) Fix/clean up running of vx tasks in NCO mode.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

Ran the `fundamental` tests on Hera with intel (see below); all passed.
```
MET_verification
MET_verification_only_vx
community_ensemble_2mems
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
MET_ensemble_verification
MET_ensemble_verification_only_vx
community_ensemble_2mems_stoch
pregen_grid_orog_sfc_climo
```

## DOCUMENTATION:
<!-- If this PR is contributing new capabilities that need to be documented, please also include updates to the RST files (docs/UsersGuide/source) as supporting material. -->
Documentation is needed but will be added once the whole set of vx PRs has been merged (see Issue #630).

## ISSUE:
Resolves #382 

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

gsketefian committed:
Change vx task names to better match final ones we want.

gsketefian committed:
Ensure that there is a vx task entry defined for each vx task and that each task has its own values for NNODES, PPN, MEM, WTIME, MAXTRIES.

gsketefian committed:
Specify location of staged forecast output for vx in the machine file for Hera and modify run_WE2E_tests.py so that this location is obtained in a platform-independent way.

gsketefian committed:
Merge branch 'develop' into feature/vx_stage_fcst_by_platform

gsketefian committed:
Undo unnecessary changes in a previous commit.

gsketefian committed:
Add experiment variable TEST_VX_FCST_INPUT_BASEDIR that specifies location of staged forecast files; modify code to correctly set VX_FCST_INPUT_BASEDIR in WE2E experiments.

gsketefian committed:
Change condition for setting default value of VX_FCST_INPUT_BASEDIR so that it is exactly consistent with the one used in run_WE2E_tests.py.

gsketefian committed:
Merge branch 'develop' into feature/vx_stage_fcst_by_platform

gsketefian committed:
Modify the way the metplus log files are named to be in line with what the DTC team wants.

gsketefian committed:
Merge branch 'develop' into feature/rename_vx_tasks

gsketefian committed:
Spacing edits.

gsketefian committed:
Merge branch 'develop' into feature/rename_vx_tasks

gsketefian committed:
Change names of METplus log files so that tool names are camel case.

gsketefian committed:
Bring in changes from another branch for splitting of SFC and UPA tasks; Add ensemble member name (if running ensemble vx) to the names of the vx tasks in the ROCOTO xml.

gsketefian committed:
Rename ex-scripts for vx tasks so that they're all lower case (as required by NCO); also rename vx task top-level sections in config_defaults.yaml so that they're all lower case (to match names of ex-scripts).

gsketefian committed:
Remove extra slashes before ${SLASH_ENSMEM_SUBDIR}.

gsketefian committed:
Rename METplus configuration files for surface and upper air.

gsketefian committed:
Change names of METplus conf and log files to follow a more consistent convention; fix bugs in the two ex-scripts exregional_run_met_pointstat_ens[mean|prob].sh that call METplus twice.

gsketefian committed:
Merge branch 'develop' into feature/rename_vx_tasks

gsketefian committed:
Allow some of the EnsembleStat tasks to use staged ensemble forecast output.  Details below.

In the ex-scripts for EnsembleStat: (1) generate a new variable FCST_INPUT_FN_TEMPLATE that contains a list of forecast input template files for each of the ensemble members and use this in the METplus configuration files, and (2) change the input base directory (INPUT_BASE) to use VX_FCST_INPUT_BASEDIR, which depending on whether forecasts are being run is set either to the location of the UPP output from the forecasts or to the staged forecast UPP output (this allows staged ensemble staged files to be used to perform ensemble vx).

In config_defaults.yaml, introduce new template variables (FCST_SUBDIR_TEMPLATE, FCST_FN_TEMPLATE, and FCST_FN_METPROC_TEMPLATE) that are/will be used in forming the variable FCST_INPUT_FN_TEMPLATE in the EnsembleStat ex-scripts.

gsketefian committed:
In the ex-scripts for deterministic PointStat verification and EnsmebleStat for point-obs, ensure that METplus is called only once and for the correct variable (instead of for both SFC and UPA).

gsketefian committed:
Fix for when the value in a bash script input key-value pair contains an equal sign.

gsketefian committed:
Add a bash utility script for parsing a METplus time-string template.  This should later be replaced with a python version.

gsketefian committed:
Add a bash utility script for parsing a METplus time-string template.  This should later be replaced with a python version.

gsketefian committed:
Add METplus variable LOG_DIR to common.conf since it is used in all METplus tasks.

gsketefian committed:
Merge commit '561cfa8b4a74' into feature/split_pcpcombine

gsketefian committed:
Remove debugging code.

gsketefian committed:
Clean up METplus conf file for APCP01h to make it more similar to the shell version.

gsketefian committed:
In the output of the GridStat tool for APCP01h, add to the MODEL variable the ensemble member name (if doing an ensemble forecast).  This makes it easier in METviewer to identify curves corresponding to different members.

gsketefian committed:
Replace GET_OBS and TN_VX with the more descriptive variables GET_OBS_LOCAL_MODULE_FN and VX_LOCAL_MODULE_FN; fix spacing.

gsketefian committed:
Include changes inadvertantly left out of previous commit.

gsketefian committed:
Include more changes inadvertantly left out of previous commit.

gsketefian committed:
Split calls to METplus's PcpCombine tool out of the tasks that run deterministic calls to the GridStat tool.  Details below.

1) Create new PcpCombine tasks (both for obs and forecasts) in the ROCOTO template XML.
2) Create new METplus configuration files for the PcpCombine tasks.
3) Disable PcpCombine from the METplus configuration files for the GridStat tool running on APCP 03-hour, 06-hour, and 24-hour forecast data.
4) In the ROCOTO template XML, add dependencies in the deterministic GridStat APCP 03, 06, and 24 hour tasks on the PcpCombine tasks.
5) Add PcpCombine task sections to config_defaults.sh.
6) Add several new vx experiment variables to config_defaults.sh that are needed to run the new PcpCombine tasks:
  a) RUN_TASKS_METVX_DET:
     This flag is to enable the deterministic verification tasks in the workflow.  If using a single deterministic forecast (either generated by running a forecast or using staged output), setting this flag to True will cause deterministic verification to be run on the post-processed forecast files.  If using an ensemble forecast (again either generated by running an ensemble forecast or using staged output), setting this flag to True will cause deterministic verification to be run on the post-processed forecast files of each member of the ensemble.
  b) RUN_TASKS_METVX_ENS:
     This flag is to enable the ensemble verification tasks in the workflow.  It is relevant only if ensemble forecast files are availabe (either generated by running an ensemble forecast or using staged output).  In this case, setting this flag to True will cause ensemble verification to be run on the ensemble as a whole (not on each member of the ensemble).
  c) VX_FIELDS:
     This is a list of the fields or groups of fields for which vx will be run.
  d) VX_APCP_ACCUMS_HH:
     If VX_FIELDS includes "APCP" (accumulated precipitation), this specifies the accumulation periods (in double-digit hours) to consider.
  e) VX_OUTPUT_BASEDIR:
     The base directory under which output from vx tasks will be placed.
  f) NUM_MISSING_OBS_FILES_MAX:
     The maximum number of missing observation files to allow before assuming that there aren't enough to perform a meaningful verification.  Note that METplus can still perform verification even when some obs files are missing, but we don't want too many to be missing.  This is a crude way of controlling how much obs data is allowed to be missing.
  g) OBS_CCPA_APCP01h_FN_TEMPLATE:
     File name template used to obtain the input observation files (in the PcpCombine_obs tasks) that contain the 01-hour APCP from which APCP for longer accumulations will be generated.
  h) OBS_CCPA_APCPgt01h_FN_TEMPLATE:
     File name template used to generate the obs files (in the PcpCombine_obs tasks) containing APCP for longer than 01-hour accumulation periods.
  Note that many of these variables will also be used in later PRs in vx tasks other than the ones that use METplus's PcpCombine tool.

gsketefian committed:
Add constant SECS_PER_HOUR needed by the script eval_METplus_timestr_tmpl.sh; modify this script so that it still works when SECS_PER_HOUR is set to a float (i.e. to 3600.0 instead of 3600).

gsketefian committed:
Activate the deterministic tasks in the deterministic vx WE2E tasks.  This is done by setting RUN_TASKS_METVX_DET to True.  This must be done due to changes to the workflow in previous commits.

gsketefian committed:
Add file inadvertantly left out of previous commit.

gsketefian committed:
Formatting changes and typo fixes in EnsembleStat_APCP03h.conf.

gsketefian committed:
Clean up comments; remove LOG_DIR from conf file since it already appears in common.conf.

gsketefian committed:
More fixes to spacing and comments.

gsketefian committed:
Add the new experiment ENS_TIME_LAG_HRS to config_defaults.sh that contains the time-lagging of each ensemble member (in integer hours).  This is needed in order for the new separate PcpCombine_fcst tasks to work properly.

gsketefian committed:
Modify the way the METplus configuration variable OBS_GRID_STAT_INPUT_DIR is set to that it works for both deterministic and ensemble verification.

gsketefian committed:
Activate both deterministic (for each ensemble member) and ensemble verification in the WE2E test MET_ensemble_verification.  This is needed because of the introduction of the new flags RUN_TASKS_METVX_DET and RUN_TASKS_METVX_ENS in a recent commit.

gsketefian committed:
Strip out the PcpCombine tool out of the METplus config files EnsembleStat_APCP[03|06|24].conf and ensure that the modified conf files are compatile with the new separate PcpCombine_[obs|fcst] workflow tasks.

gsketefian committed:
Modify the METplus conf files GridStat_ensmean_APCP[03|06|24].conf to ensure that the run_MET_GridStat_vx_ensmean_APCP[03|06|24] workflow tasks are compatible with the new separate PcpCombine_[obs|fcst] workflow tasks.

gsketefian committed:
Re-introduce the ENS_VAR1_THRESH and accompanying variables to EnsembleStat_APCP[03|06|24]h.conf since without these, the ensemble probabilities (frequencies) needed by the downstream "...ensprob..." tasks are not included in the GridStat netCDF output files.

gsketefian committed:
Modify the METplus conf files GridStat_ensprob_APCP[03|06|24].conf to ensure that the run_MET_GridStat_vx_ensprob_APCP[03|06|24] workflow tasks are compatible with the updated run_MET_EnsembleStat_vx_APCP[03|06|24]h and new separate PcpCombine_[obs|fcst] workflow tasks.

gsketefian committed:
Fixes to comments in EnsembleStat_APCP01h.conf so that it is more similar to the EnsembleStat_APCP[03|06|24]h.conf files.

gsketefian committed:
Merge branch 'develop' into feature/split_pcpcombine

gsketefian committed:
Remove construction and use of the variable FCST_INPUT_FN_TEMPLATE in the workflow tasks that run EnsembleStat since this procedure is only needed when doing time lagging (which is not yet in the App).

gsketefian committed:
For all deterministic vx tasks that don't already have it, add the member name to the end of the MODEL variable's value.

gsketefian committed:
Undo all cosmetic changes to GridStat_APCP01h.conf since that's better done as part of a different PR.

gsketefian committed:
Undo all cosmetic changes to EnsembleStat_APCP01h.conf since that's better done as part of a different PR.

gsketefian committed:
Changes left out of previous commit.

gsketefian committed:
Changes left out of previous commit.

gsketefian committed:
Move CDATE portion of input path to EnsembleStat tasks from METplus conf files back to the ex-scripts.  This change should be part of another PR.

gsketefian committed:
Changes left out of previous commit.

gsketefian committed:
Changes left out of previous commit.

gsketefian committed:
Add auxiliary functions needed in verification.

gsketefian committed:
Edit comments.

gsketefian committed:
Rearrange variable location in config_defaults.sh.

gsketefian committed:
For consistency with names of other entities, drop the "_apcp" from the section name of the PcpCombine tasks in config_defaults.sh; make edits to comments.

gsketefian committed:
Clean up comments in PcpCombine METplus conf files.

gsketefian committed:
Add j-job for PcpCombine tasks (inadvertantly left out of previous commits).

gsketefian committed:
Clean up comments.

gsketefian committed:
Remove unused variable TN from PcpCombine tasks in the ROCOTO XML; use bash variable expansion/substitution operators to get argument name and value in arg=value pairs; clean up comments.

gsketefian committed:
Add j-job for running Pb2nc tasks.

gsketefian committed:
Add a METplus configuration file for the Pb2nc tool.

gsketefian committed:
Add ex-script for running Pb2nc tasks.

gsketefian committed:
Split out calls to Pb2nc tool into separate tasks; a lot of other cleanup (will document in PR message).

gsketefian committed:
Merge branch 'develop' into feature/split_pb2nc

gsketefian committed:
Merge branch 'develop' into feature/split_pb2nc

gsketefian committed:
Address PR comments.

gsketefian committed:
Jinja templates for GridStat running on APCP01h, REFC, and RETOP.

gsketefian committed:
Changes to ROCOTO xml to get templated versions of GridStat tasks for APCP01, REFC, and RETOP to work.

gsketefian committed:
Remove trailing whitespace.

gsketefian committed:
Add MRMS obs file name templates.

gsketefian committed:
Bug fixes.

gsketefian committed:
Modify deterministic PointStat tasks in ROCOTO XML and the corresponding configuration files so that the files are Jinja templates.

gsketefian committed:
Clean up commented out code.

gsketefian committed:
In ROCOTO template XML, modify GridStat tasks for APCP > 01 hour so that the tasks work with Jinja template METplus conf file; remove unneeded variables from other tasks.

gsketefian committed:
Minor edits to comments and spacing.

gsketefian committed:
Modify PointStat_ensmean tasks for SFC and UPA in ROCOTO XML and corresponding METplus conf files to allow use of templates.

gsketefian committed:
Modify GridStat_ensmean tasks for APCP 01, 03, 06, and 24 hours such that they use templates for the METplus config files.

gsketefian committed:
Modify the EnsembleStat METplus config files for APCP so that the names of the output files are consistent with what the GridStat_ensmean and PointStat_ensmean tasks expect to read in.

gsketefian committed:
Change the METplus config file for the GridStat_ensmean task for REFC into a jinja template although it will not be used.

gsketefian committed:
Add J-jobs for ensemble mean and probabilistic GridStat and PointStat tasks.

gsketefian committed:
Modify ROCOTO XML and conf files to have GridStat_ensprob tasks for all four APCP variables (01, 03, 06, and 24 hour accums) use templates for the METplus conf files.

gsketefian committed:
Modify GridStat_enspoint tasks for REFC and RETOP so they can use Jinja template METplus conf files.

gsketefian committed:
Change METplus config files for GridStat_ensprob tasks so that they're Jinja templates.

gsketefian committed:
Modificataions to allow PointStat_ensprob tasks for SFC and UPA to use templates from which to generate METplus config files.

gsketefian committed:
Minor edits to comments.

gsketefian committed:
Remove files that are now replaced by GridStat_ensmean_APCPgt01h.conf.

gsketefian committed:
Change the METplus config file for the GridStat_ensmean task for RETOP into a jinja template although it is currently not used in the workflow.

gsketefian committed:
Minor edits to comments.

gsketefian committed:
Add j-job and ex-script for running EnsembleStat for APCP, all accumulations (01, 03, 06, and 24); modify METplus conf file for accumulation of 01 hour so it's a template; add template Metplus conf file for accumulatons > 01hour; modify ROCOTO xml to run with new j-job.

gsketefian committed:
Minor bug fixes.

gsketefian committed:
Bug fixes to get EnsembleStat to output all the fields required by downstream tasks.

gsketefian committed:
Edits to comments.

gsketefian committed:
Change METplus conf files for EnsembleStat operating on REFC and RETOP to templates and add necessary variables to correspondig ROCOTO template XML tasks.

gsketefian committed:
Fix up comments and spacing.

gsketefian committed:
Change METplus conf files for EnsembleStat operating on SFC and UPA to templates and add necessary variables to the correspondig ROCOTO template XML tasks.

gsketefian committed:
Edits to comments.

gsketefian committed:
Merge branch 'develop' into feature/METplus_templates

gsketefian committed:
Delete unneeded METplus config files (replaced by generic "_gt01h" version).

gsketefian committed:
Minor edits for consistency between METplus config files.

gsketefian committed:
Remove unneeded ex-scripts.

gsketefian committed:
Remove unneeded j-jobs.

gsketefian committed:
Add files left out of previous commits.

gsketefian commented:
@danielabdi-noaa This PR doesn't yet work in NCO mode, but I'm working on that now and will hopefully get it done this afternoon.

@christinaholtNOAA This is the PR I was referring to on Friday.  It has only a few minor (and duplicate) mods to the vx tasks in `FV3LAM_wflow.xml`.  Not sure where you're at with your PR.  If you get your PR in before mine, that's fine, I will incorporate the xml changes afterward.  Also, please let me and @MichaelLueken know if you'd like us to wait after yours is in before merging this PR; that should be fine.

@JeffBeck-NOAA @linlin-pan @willmayfield @michelleharrold In case you have time to review, that would be great.  Thanks.

gsketefian committed:
Remove LOG_SUFFIX from vx ex-scripts since it is no longer used in the METplus conf files.

JeffBeck-NOAA reviewed (approved):
Approved pending any necessary NCO mode changes.

gsketefian committed:
Modifications to get NCO mode to work with staged files.

mkavulich commented on jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT:
```diff
@@ -52,18 +51,40 @@ print_info_msg "
+valid_vals_MET_TOOL=( "GENENSPROD" "ENSEMBLESTAT" )
+check_var_valid_value "MET_TOOL" "valid_vals_MET_TOOL"
+
+if [ "${MET_TOOL}" = "GENENSPROD" ]; then
```
All of this `_sc` vs `_pc` logic seems very complicated for something that doesn't seem to influence the code in any way; unless I'm missing something the "snake case" variables are only ever printed to log files. Is there some other reason I am missing?

gsketefian replied:
Yeah, it's left over from the time when some things were named "grid_stat_..." and others used "GridStat_...".  Let me take a look and see if I can simplify.

gsketefian replied:
@mkavulich I cleaned up the MET/METplus naming code, hopefully to your approval.  I removed the code in the J-jobs, and now the ex-scripts don't need arguments and get the variable `MET_TOOL` directly from the environment.  I added a function (`get_met_metplus_tool_name.sh`) that sets the names of the MET and METplus tools.  I want to keep both names still because they are both currently used; maybe we'll simplify the ex-scripts later and use only one of the names to refer to the tool, but let's leave that for another PR.  Also, I know we want to do all python scripts, but I haven't gotten a chance to do that yet (will do it in a future PR if someone else doesn't get to it before).  Let me know what you think.

Also, I reran the WE2E tests `MET_verification_only_vx` and `MET_ensemble_verification_only_vx`, and the both passed.

gsketefian committed:
Further fixes to make vx tasks work in NCO mode; add WE2E test for ensemble verification with staged forecasts.

gsketefian committed:
Bug fix.

gsketefian committed:
Add MET_ensemble_verification_only_vx test to set of fundamental tests on Hera.  This test runs ensemble vx from staged forecast and obs.

gsketefian commented:
@danielabdi-noaa @christinaholtNOAA @MichaelLueken @JeffBeck-NOAA Here's an update:
1) I fixed (and tested) NCO mode, in the process cleaning up where the vx output gets placed (output from some tasks were (I think inadvertantly) being placed in redundant CDATE-based subdirectories under `$COMOUT`).
2) I intentionally did not make any big changes to `FV3LAM_wflow.xml` to keep it easier to integrate into the new XML generation system of @christinaholtNOAA.
3) I retested all the fundamental tests on Hera.  I also ran the four vx tests (`MET_verification`, `MET_verification_only_vx`, `MET_ensemble_verification`, and `MET_ensemble_verification_only_vx`) in both `community` and `nco` modes, and they all passed.

So this PR is ready for the automated tests (and needs another approval).  Thanks.

gsketefian commented:
@danielabdi-noaa @JeffBeck-NOAA @MichaelLueken If you're interested in the layout of the vx output in NCO mode, you can look on Hera under the directory
```
/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/DTC_ensemble_task_pythonize_nco/expt_dirs
```
in the subdirectories
```
nco_dirs.works01.MET_ensemble_verification_nco
nco_dirs.works01.MET_ensemble_verification_only_vx_nco
nco_dirs.works01.MET_verification_nco
nco_dirs.works01.MET_verification_only_vx_nco
```
Under these subdirectories, look in
```
com/rrfs/v1.0.0/rrfs.20190615/00/metout/
```

danielabdi-noaa reviewed (approved):
Thanks @gsketefian. Looks good to me!

MichaelLueken commented on parm/metplus/Pb2nc_obs.conf:
```diff
@@ -115,7 +115,7 @@ OUTPUT_BASE = {{output_base}}
 [filename_templates]
 #
-# Template to look for prepbvur input to PB2NC relative to PB2NC_INPUT_DIR.
+# Template for prepbvur input to PB2NC relative to PB2NC_INPUT_DIR.
```
Should prepbvur be prepbufr?

gsketefian replied:
Yep!  Fixed it.

MichaelLueken reviewed (approved):
@gsketefian These changes look good to me!  I have noted a minor spelling issue in `parm/metplus/Pb2nc_obs.conf` (prepbuvr rather than prepbufr), which I have gone ahead and noted in my review.

Will go ahead and approve and submit automated Jenkins testing now.  Please note that Gaea is still down, so it might take some time for the tests to fully complete.

MichaelLueken commented:
@gsketefian All tests have successfully passed, with the exception of Gaea, which is still down.

gsketefian committed:
Fix type to address Mike L's PR comments.

mkavulich commented:
@gsketefian When I tried running the verification WE2E tests on Hera, they all showed failures in GridStat tasks for ACPC. Here is an example log file: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/expt_dirs/community/MET_verification/log/run_MET_GridStat_vx_APCP03h_2019061500.log

mkavulich reviewed (changes requested):
@gsketefian When I tried running the verification WE2E tests on Hera, they all showed failures in GridStat tasks for ACPC. Here is an example log file: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/expt_dirs/community2/MET_verification_only_vx/log/run_MET_GridStat_vx_APCP03h_2019061500.log

mkavulich commented:
Sorry about all the comments; I accidentally hit "submit" earlier on an earlier comment but I wanted to get my note about the failures before this was merged.

To go into further detail, I'm not sure why my tests are failing while yours seem to be succeeding. I started with a fresh clone of your branch on Hera (/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/ufs-srweather-app/). It seems to be specifically the APCP03 and 06 GridStat tasks that are failing: the APCP01 GridStat tasks are successful.

@MichaelLueken Can you point me to the location on disk of the Jenkins test on Hera so I can see what the successful log files look like?

gsketefian committed:
Add missing file.

gsketefian commented:
@mkavulich It failed because I had neglected to commit a new conf file (that I have on disk, so my tests succeed).  I just added it and hopefully it will work.

@MichaelLueken I'm also curious how the tests succeeded without that file...

gsketefian committed:
Introduce a new function (set_met_tool_name.sh) that sets the names of the MET/METplus tools; remove "met_tool_[sc|pc]" arguments to the vx ex-scripts and directly use the environment variable MET_TOOL along with the new function (set_met_tool_name) to set the MET/METplus tool names needed in the vx tasks.

gsketefian committed:
Rename MET/METplus name-related variables for clarity.

gsketefian commented:
@mkavulich Did your manual NCO-mode tests on Hera succeed?

MichaelLueken commented:
@mkavulich and @gsketefian The Jenkins tests don't run on Hera currently.  Currently, automated tests only run on Cheyenne, Jet, Gaea, and Orion.  While new allocations are in place for Hera, Jet, and Gaea, I'm unsure whether these new allocations have been propagated through Jenkins.

mkavulich commented:
@gsketefian The tests on Hera have passed on your latest changes. But RE: my comments about MET tool names, I don't think you understood my point...I was looking to simplify some logic, but your latest commits have only *increased* the complexity of the logic. It's clever, sure, but it seems like completely unnecessary complexity. Why are we maintaining a whole bunch of logic (and now even a special function) to maintain two different variable names for each MET tool when one of the names is never even used except for in print statements? ush/get_met_metplus_tool_name.sh is 124 lines that does nothing but take one string and return two different strings....but one of those strings (`met_tool_name`) never gets used except in print statements! Can we not just use that second string (`metplus_tool_name`) in print statements and greatly simplify the logic? 

Again, maybe there's a reason these additional variables are needed that I'm missing, but I don't see it.

gsketefian commented:
@mkavulich I'm anticipating that we may need to refer to MET vs. METplus versions of the tool in the ex-scripts at some point.  Before this PR, both were used in various places.  I went through all the config files and converted as much of the MET tool references to METplus references.  I'm getting to what you're suggesting (getting rid of all MET tool references), but that's something we should discuss in the DTC vx group.  We may at some point need to call MET without going through METplus.  I tried to reduce everything to that one function to still keep the capability but have it all in one place (so it's easy to remove in the future if we want).  I prefer to keep it and remove after a discussion, but if you feel very strongly then I'll remove it.

mkavulich commented:
@gsketefian I'm fine with that reasoning, just wanted to make sure we weren't incurring technical debt without a reason.

JeffBeck-NOAA commented:
@MichaelLueken, I think this PR is ready.  The only problem is that @gsketefian is on leave this week and there is a minor conflict with the hera.yaml machine file (change in default paths).  Are you comfortable with me resolving this conflict and then proceeding with the merge?

MichaelLueken commented:
@JeffBeck-NOAA I certainly don't mind if you went ahead and corrected the conflict.  The issue is that PRs that change the FV3LAM_wflow.xml file are on hold until @christinaholtNOAA's PR #676 has been merged.

@christinaholtNOAA Since both this PR and @chan-hoo's PR #682 have been approved and automated tests have been run and passed, can we move forward with these two PRs?  Unfortunately, it would mean needing to apply these two updates to your PR before your work can be merged.  Thank you very much for your time and consideration.

chan-hoo commented:
@MichaelLueken, please merge my PR https://github.com/ufs-community/ufs-srweather-app/pull/682 before the PR https://github.com/ufs-community/ufs-srweather-app/pull/676 if it passed the tests :) I am going to create the `production/aqmv7` branch for production of AQMv7 once my PR is merged. It (without the huge change) will make my life much easier :)

MichaelLueken commented:
@JeffBeck-NOAA Speaking with @christinaholtNOAA, the final changes should be in place and we are hoping that the Jenkins tests will be submitted today on PR #676.  So long as there are no failures with the Jenkins tests and manual testing on Jet, then her PR will be merged either tomorrow or first thing Friday.  If there are issues, then I will be able to move forward with the changes in this PR.  Thanks!

JeffBeck-NOAA committed:
Merge branch 'develop' into feature/METplus_templates

JeffBeck-NOAA commented:
@MichaelLueken, thanks.  I've resolved the conflicts for this PR and merged the commit.

mkavulich committed:
Updates feature/METplus_templates to current develop (#1)

* [develop] Adds a YAML interface for creating a Rocoto XML. (#676)

Refactors the creation of a Rocoto XML to use a very generic Jinja2 template that is flexible enough to meet the needs of various workflow configurations supported by SRW. Specifically, it allows for a completely arbitrary workflow to be created under SRW, which includes the addition of completely arbitrary tasks on top of the predefined ones.

---------

Co-authored-by: Michael Kavulich <kavulich@ucar.edu>

* [develop] Change the build log output file extension from log to txt (#690)

When pipeline files are archived to s3 bucket, retrieving the file via a browser attempts to render/display files of known extensions. A browser doesn't generally understand what to do with a .log extension (e.g. build.log). For ease of use in the CI Dashboard, which is a static HTML page, the s3 archived build log needs a .txt extension (e.g. build.txt).

* Add "MET_TOOL" definitions to new XML definition YAMLs

* Fix incorrect YAML if block in config_defaults, remove non-needed "USCORE_ENSMEM_NAME_OR_NULL" variable

* - Convert new test "MET_ensemble_verification_only_vx" to new YAML format
 - Fix f-string for utils.py error message

* Fixing more failures (still more to go)

* More fixes, got stand-alone verification test to pass!

 - Fix copy-paste errors in parm/workflow yamls
 - Update corrected variables for new names in exscripts

* Improvement for monitor jobs script: if in debug mode, print the number of tasks that succeeded and failed for failed experiments

* Forgot to include VX_FCST_INPUT_DIR definition for MET_ensemble_verification_only_vx test

* Correct script for task_run_MET_EnsembleStat_vx_APCP

* Pull out CATE and ENSMEM_INDEX from default VX_FCST_INPUT_DIR. My naive attempt to simplify things was the root of all my problems!

* Everything working! Just need to solve problem of non-existent metatask dependencies!

* Fix last failing ensemble test, fundamental tests and all verification tests now pass on Hera!

---------

Co-authored-by: Christina Holt <56881914+christinaholtNOAA@users.noreply.github.com>
Co-authored-by: Bruce Kropp - Raytheon <104453151+BruceKropp-Raytheon@users.noreply.github.com>

gsketefian committed:
For clarity, rename VX_FCST_INPUT_DIR to VX_FCST_INPUT_BASEDIR.

gsketefian committed:
Remove "mode=serial" attribute of metatasks since that causes a task to be launched only after the previous task has completed even when there is no dependency of the task on the previous one.

gsketefian committed:
The PoinStat tasks need to depend on the Pb2nc task because METplus reads in the netcdf versions of the NDAS obs.

gsketefian committed:
For the WE2E test MET_verification_only_vx, in the "and" clause of the dependency section of the PointStat tasks, empty out the dependencies on the post and get_obs tasks since those are not run for this test.

gsketefian committed:
Reintroduce the variable FCST_SUBDIR_TEMPLATE because it is needed when performing time-lagging vx.

gsketefian committed:
Remove variable SLASH_ENSMEM_SUBDIR_OR_NULL from default vx task and place it only ub tasks that need it; define SLASH_ENSMEM_SUBDIR_OR_NULL locally in a loop when we loop over members in EnsembleStat vx task.

gsketefian committed:
For the MET_ensemble_verification_only_vx task, remove dependency of PointStat task on get_obs_... and do_post tasks.

gsketefian committed:
Bug fix.

gsketefian committed:
Define ENSMEM_INDX only in those vx tasks that need it (i.e. those that are looped over ensemble members).

gsketefian committed:
For the Pb2nc obs task, remove the (undefined) ensemble member name/number from name of METplus config and task log files.

gsketefian committed:
Merge branch 'develop' into feature/METplus_templates

gsketefian committed:
Bug fixes due to inadvertently leaving out some aspects of vx in the METplus conf files (MLCAPE and P-type).

gsketefian commented:
@MichaelLueken I updated this PR and reran (on Hera) the four verification tests:
```
MET_verification_only_vx
MET_ensemble_verification_only_vx
MET_verification
MET_ensemble_verification
```
In addition, I ran the fundamental tests, which currently are:
```
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16
```
All tests were successful.  I think this PR is ready for the Jenkins tests.

gsketefian commented:
@MichaelLueken I see the error message "This commit cannot be buiild".  I'd like the see the logs, but clicking on "Details" doesn't take me there (browser hangs).  Are the logs not public?

MichaelLueken commented:
@gsketefian Unfortunately, the Jenkins logs are no longer public due to security concerns.  I'll look through the failures now, but the due to the upgrading of PrgEnv on Gaea this week, Gaea is having issues.  I'll let you know if there are any failures that will need to be addressed before this work can be merged.

MichaelLueken commented:
@gsketefian 

On Cheyenne GNU, the `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` test failed intermittently in the post.  This is an ongoing issue that has been seen in all PRs since the test was added to the coverage suite and might be due to issue #652.  This will not keep your PR from being merged.

On Hera GNU, both the `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` and `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` tests failed.  The `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` test has been failing since it was introduced into the coverage tests.  A rerun of the `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` test showed that it passes.  Can safely move forward.

On Hera Intel, the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test failed in the run_fcst task.  This is known issue (#731).

Since the failures are expected and the rerun of the one unexpected failure successfully passed, I will now move forward with merging this work.

gsketefian commented:
@MichaelLueken Great, thank you!
